src/test/controller/UserController.test.ts

```diff
@@ -399,4 +375,59 @@ describe('Unit tests for UserController', function () {
 
     expect(actualUser).toMatchObject(expectedObject);
   }
+
+  /**
+   * Save the promotion and user to the database. Also add the saved promotion into the database
+   * * Set transaction level to serializable and set lock level to pessimistic write to avoid deadlocks
+   * */
+  async function addSavedPromotion(user: User, promotion?: Promotion) {
```

Comment:
I think the deadlock may be occurring because we're trying to do too much in one function. In this case, we should have a function specifically for creating a user, and another function for creating a promotion and a saved promotion. In the second function, reading the user would not require a lock, and we can simply abort the transaction if the user gets deleted during this time.

Reply:
Hmm I'm not too sure I understand how moving it to separate functions will solve the deadlock. Unless you mean one transaction for saving the user and another transaction for creating a promotion and a saved promotion.

Right now there doesn't seem to be any deadlocks for saving a user, promotion, and saved promotion. The deadlock seems to happen when we send the request to the controller after saving these entities.